rustls/src/client/hs.rs

```diff
@@ -271,6 +271,56 @@ fn emit_client_hello_for_retry(
     // Do we have a SessionID or ticket cached for this host?
     let tls13_session = prepare_resumption(&input.resuming, &mut exts, suite, cx, config);
 
+    // Extensions MAY be randomized
+    // but they also need to keep the same order as the the previous ClientHello
+    exts.sort_by_cached_key(|new_ext| {
+        // PSK extension is always last
+        if let ClientExtension::PresharedKey(..) = new_ext {
+            return usize::MAX;
+        }
+
+        if !input.hello.sent_extensions.is_empty() {
+            // There is an existing saved ClientHello, we make some assumptions here:
+            // - Extensions are unique (mandated by RFC 8446 section-4.2)
+            // - If extensions are removed, all ordering between the existing extension relative to
+            // the old ClientHello is kept, this is unclear from the RFC but it is the safer
+            // approach for compatibility with assumptions in existing TLS server implementations
+            // - If extensions are added they are randomized after the list of existing extensions
+            // and before PSK extension. The existing extensions are kept in the same order. Again
+            // this is unclear from the RFC but it is the cautious approach to maximize
+            // compatibility, since shuffling is not explicitly mandated.
+            match input
+                .hello
+                .sent_extensions
+                .iter()
+                .enumerate()
+                .find(|(_, &saved_ext)| saved_ext == new_ext.ext_type())
+            {
+                Some((pos, _)) => pos,
+                _ => match crate::rand::random_u32(config.provider.secure_random) {
+                    Ok(value) => {
+                        // We do not know exactly how many extensions have been sorted as we
+                        // can remove extensions (e.g. early_data) but we know that if no extensions
+                        // have been removed we want to put the same ones in the same
+                        // positions so worst case we need to skip `sent_extensions.len()` elements
+                        // This will not create any issues as this number is not an actual index of
+                        // the vector but just set the numerical ordering used for the extensions
+                        (value as usize % (usize::MAX - input.hello.sent_extensions.len()))
+                            + input.hello.sent_extensions.len()
```

Comment:
I don't think it's significant in this context, but it feels worth noting that this isn't uniformly random. As I understand things it will suffer a [modulo bias](https://romailler.ch/2020/07/28/crypto-modulo_bias_guide/).

Maybe worth mentioning that explicitly in the comment?

Reply:
I was looking for that too, but I thought it was mitigated? IIRC mitigating it is not too hard.